Add refactored filter and config controller for API Priority and Fairness

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:
This adds the request filter and its associated local configuration controller for tthe API Priority and Fairness feature.  This is an alternative to #85319 that manages the top level of configuration data by locking rather than sync/atomic.Value storage of immutable objects.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

**Special notes for your reviewer**:
This needs a release note but it will be supplied in a future PR.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP] https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
```

MikeSpreitzer commented:
/sig api-machinery

MikeSpreitzer commented:
/cc @yue9944882 
/cc @aaron-prindle 
/cc @lavalamp 
/cc @deads2k 
/cc @jktomer

MikeSpreitzer commented:
@mars1024 
@bjrara

MikeSpreitzer commented:
The force-push to b64b7ee76fc fixes gofmt and golint issues.

fedebongio commented:
/assign @lavalamp @jktomer

bjrara commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -553,6 +559,16 @@ func BuildAuthorizer(s *options.ServerRunOptions, versionedInformers clientgoinf
 }
 
+// BuildRequestManager constructs the request manager
+func BuildRequestManager(s *options.ServerRunOptions, extclient clientgoclientset.Interface, versionedInformer clientgoinformers.SharedInformerFactory) utilflowcontrol.Interface {
```
as far as i can remember, in previous reviews, yue suggested to abort requestManager concept here as it was split into filter and controller. is that change going to be in another pr?

MikeSpreitzer replied:
I am not sure I understand your words.  Are you raising an issue about the name, or about how changes are packaged into PRs, or something else?

This PR includes both the filter and the configuration controller.  Particularly after the recent refactorings, there is very little in the filter; almost all the action is in the configuration controller.

Yes, "request manager" is an outdated term.  I plan to submit a PR in the future to replace it.

bjrara replied:
it's about the name. just to make sure that this name is going to be aborted. thanks for answering.

MikeSpreitzer replied:
OK.
And this name appears only in this PR, so I will fix it in this PR.

MikeSpreitzer commented:
The last two force-pushes fix tiny things found by automated nit picking.

MikeSpreitzer commented:
The force-push to bae190ceffb fixed a gofmt oversight.

MikeSpreitzer committed:
Added API Priority and Fairness filter and config consumer

MikeSpreitzer commented:
The force-push to b13168f15dc is a rebase onto master.

MikeSpreitzer commented:
The force-push to 1090716a100 extends the rebase onto master with updating bazel and vendor.

MikeSpreitzer commented:
And the force-push to  67560f52689 adds missing BUILD files.

MikeSpreitzer committed:
Cleanup in queueset.go

Added explanation of naming convention for methods of the queueset
type.

Added missing "Locked" suffix to a couple of method names.

Removed overlooked vestige of EmptyHandler.

Removed unneeded handling of bad queue indices.

Removed outdated comment about not yet implementing concurrency
limiting without queueing.

MikeSpreitzer committed:
Revise test infrastructure for behavioral tests

Updated the generation of FlowSchemas to be able to provide uniquely
matching and comprehensively mismatching digests, so that it will be
easier to identify correct behavior.

MikeSpreitzer commented:
The force-push to 601911cf31f fixes lint and staticcheck nits.

MikeSpreitzer committed:
A little more prep for behavioral testing of the config consuming controller

yue9944882 commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -523,6 +525,10 @@ func buildGenericConfig(
 		lastErr = fmt.Errorf("failed to initialize admission: %v", err)
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.APIPriorityAndFairness) {
```
```suggestion
	if utilfeature.DefaultFeatureGate.Enabled(genericfeatures.APIPriorityAndFairness) {
		_, flowcontrolAPIEnabled := s.APIEnablement.RuntimeConfig["flowcontrol.apiserver.k8s.io"]
		_, allAPIEnabled := s.APIEnablement.RuntimeConfig[resourceconfig.APIAll]
		_, alphaAPIEnabled := s.APIEnablement.RuntimeConfig[resourceconfig.APIAlpha]
		switch {
		case !s.GenericServerRunOptions.EnableInflightQuotaHandler:
		case !flowcontrolAPIEnabled && !alphaAPIEnabled && !allAPIEnabled:
		default:
			genericConfig.FlowControl = BuildPriorityAndFairness(s, clientgoExternalClient, versionedInformers)
		}
	}
```
there's currently 3 conditions for enabling the feature: (a) the `--enable-inflight-quota-handler` flag (b) the feature-gate (c) the API enablement. we can simply check all these conditions at this point when constructing the flowcontrol system so that we only check if the system instance is nil on deciding whether to replace the legacy handler w/ the new one.

MikeSpreitzer replied:
I do not clearly remember the reason that the `--enable-inflight-quota-handler` flag was introduced.  I have a vague memory that it was about switching the config source from max-mutating-inflight + max-readonly-inflight to a single number.

It seems silly to add a flag whose purpose is to enable the API Priority and Fairness feature.  That's what feature gates are already there for.

lavalamp replied:
> I do not clearly remember the reason that the --enable-inflight-quota-handler flag 

Possibly at my request, since the feature flag is temporary and the need to optionally not run this is permanent.

lavalamp replied:
At any rate the suggested change here should go in a different PR I think where we can debate it separately.

yue9944882 commented on pkg/kubeapiserver/server/insecure_handler.go:
```diff
@@ -20,14 +20,20 @@ import (
 func BuildInsecureHandlerChain(apiHandler http.Handler, c *server.Config) http.Handler {
 	handler := apiHandler
-	handler = genericfilters.WithMaxInFlightLimit(handler, c.MaxRequestsInFlight, c.MaxMutatingRequestsInFlight, c.LongRunningFunc)
+	if feature.DefaultFeatureGate.Enabled(features.APIPriorityAndFairness) && false {
```
it doesnt have to be protected by the feature-gate, checking if it's nil would be sufficient. (see my last comment on where the flowcontrol system is constructed)

MikeSpreitzer replied:
I chose to not check `nil` on the grounds that if the admin asked for this feature and the field is `nil` then something is wrong and the apiserver should crash.

lavalamp replied:
the `&& false` deserves a comment?

MikeSpreitzer replied:
Definitely.
It is a temporary thing I introduced for debugging during initial development.  It was motivated by a deadlock inside the code, which prevented all future requests from being handled --- including calls to GET /debug/pprof/ things.

MikeSpreitzer replied:
Added the comment.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -606,6 +612,18 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
+	requestManagerConfigControllerHookName := "apiserver/request-management-config-controller"
+	if feature.DefaultFeatureGate.Enabled(features.APIPriorityAndFairness) && !s.isPostStartHookRegistered(requestManagerConfigControllerHookName) {
+		err := s.AddPostStartHook(requestManagerConfigControllerHookName, func(context PostStartHookContext) error {
+			go c.FlowControl.Run(context.StopCh)
```
do we have to run this in the background even if somebody sets `--enable-inflight-quota-handler=false` in the flags? (but w/ the feature-gate enabled)

MikeSpreitzer replied:
Yes.  `c.FlowControl.Run` runs indefinitely, but the post-start-hook should return quickly.

jktomer replied:
I think the question is, does this have to run *at all* if `--enable-inflight-quota-handler=false`?

MikeSpreitzer replied:
My first question is, why is there a `--enable-inflight-quota-handler` flag?

jktomer replied:
The name is a bit odd to me, but it came from @yue9944882 in PR #76413; it controls the use of the new priority & fairness code (when it's off, we should be using the previous max-in-flight code).

lavalamp replied:
We've renamed the feature and no one should be using the flag yet so it's probably a good time to fix the name.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:
```diff
@@ -0,0 +1,124 @@
+
+		// Skip tracking long running requests.
+		if longRunningRequestCheck != nil && longRunningRequestCheck(r, requestInfo) {
+			klog.V(6).Infof("Serving RequestInfo=%#+v, user.Info=%#+v as longrunning\n", requestInfo, user)
```
nit: given that it will be printing on every non-long-running request. a verbosity of 8 or 10 looks better?

MikeSpreitzer replied:
Verbosity 8 is very detailed, it lists HTTP request and response bodies.  Verbosity 5 introduces per-request stuff.  So I think verbosity level 6 is good here.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,700 @@
+					}
+					idle3 := afterExec()
+					if idle3 {
+						cfgCtl.maybeReap(plName)
```
maybeReapLocked? `cfgCtl.lock` will be deadlock

MikeSpreitzer replied:
this anonymous function is also called by the client without `cfgCtl` locked.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,700 @@
+				exec, idle2, afterExec := req.Wait()
+				execID := &exec
+				if idle2 {
+					cfgCtl.maybeReap(plName)
```
maybeReapLocked? `cfgCtl.lock` will be deadlock

MikeSpreitzer replied:
`startFn` is invoked by the client, without `cfgCtl` locked.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,700 @@
+
+// Call this if both (1) plState.queues is non-nil and reported being
+// idle, and (2) cfgCtl's lock has not been released since then.
+func (cfgCtl *configController) maybeReapLocked(plName string, plState *priorityLevelState) {
```
nit: not really sure but probably `maybeReapUnlocked` is a better name?

MikeSpreitzer replied:
This file is following the same naming convention as the queueset type: suffix "Locked" means the caller has to hold the lock.  I will add a comment in this file.

MikeSpreitzer replied:
Added comment, on the configController type.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,700 @@
+			continue
+		}
+
+		concurrencyLimit := int(math.Ceil(float64(meal.cfgCtl.serverConcurrencyLimit) * float64(plState.config.Limited.AssuredConcurrencyShares) / meal.shareSum))
```
nit: better if we comment here that applying ceiling can lead to some inaccuracy but it's by-design ignored in the current phase. otherwise future-visitors can be confused..

MikeSpreitzer replied:
done

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,700 @@
+	}
+	if (plSpec.Type == fctypesv1a1.PriorityLevelEnablementExempt) != (plName == fctypesv1a1.PriorityLevelConfigurationNameExempt) {
+		// This package does not attempt to cope with a priority level dynamically switching between exempt and not.
+		return nil, errors.New("non-alignment between name and type")
```
it wont be happening (this comment also applies to L539)

https://github.com/kubernetes/kubernetes/blob/29b09c7fb1a8d38ab9df4873553cc0b8d97aae95/pkg/apis/flowcontrol/validation/validation.go#L365-L367

MikeSpreitzer replied:
I got beaten up in previous reviews about long-distance dependencies.  There seems to be the idea of not relying on the validation because that is rather distant code.

MikeSpreitzer commented on pkg/registry/flowcontrol/rest/storage_flowcontrol.go:
```diff
@@ -155,24 +155,24 @@ func ensure(flowcontrolClientSet flowcontrolclient.FlowcontrolV1alpha1Interface,
 		_, err := flowcontrolClientSet.FlowSchemas().Create(flowSchema)
 		if apierrors.IsAlreadyExists(err) {
-			klog.V(3).Infof("system preset FlowSchema %s already exists, skipping creating", flowSchema.Name)
+			klog.V(3).Infof("Suggested FlowSchema %s already exists, skipping creating", flowSchema.Name)
```
This discussion is attached to a log message rather than an error string.  Note that even the cited document capitalizes log messages.

lavalamp replied:
Yeah log messages should read normally. The error rule is so it is easier to for a thing higher in the stack to prepend information while leaving it a proper sentence.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,700 @@
+			klog.V(7).Infof("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, matchID=%p", rd, fs.Name, fs.Spec.DistinguisherMethod, plName, matchID)
+			startCalls := 0
+			startFn := func(ctx context.Context, hashValue uint64) (execute bool, afterExecution func()) {
+				startCalls++
```
oops, missing synchronization.  Let's use `sync/atomic` here.

MikeSpreitzer replied:
done

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,700 @@
+				klog.V(7).Infof("For matchID=%p, startFn(..) => exec=%v, execID=%p", matchID, exec, execID)
+				afterCalls := 0
+				return exec, func() {
+					afterCalls++
```
Missing synchronization here too.

MikeSpreitzer replied:
used sync/atomic here too

MikeSpreitzer committed:
Added commentary and synchronization based on review

jktomer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:
```diff
@@ -235,144 +213,136 @@ const (
+		go func() {
+			defer runtime.HandleCrash()
+			qs.goroutineDoneOrBlocked()
+			select {
```
I'm a bit confused. There's nothing blocking between the beginning of this goroutine and the select, so the Done channel will almost certainly not be closed and this will almost never do anything, right?

MikeSpreitzer replied:
This `select` statement has no way out except reacting to `doneCh`.  In other words, this select statement waits for doneCh to be closed.

jktomer replied:
Oh, right, there's no default.

In that case, just `_ := <-doneCh` instead of a select at all?

MikeSpreitzer replied:
done

jktomer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:
```diff
@@ -235,144 +213,136 @@ const (
+	}
+
+	// ========================================================================
+	// Step 2:
```
I'm unclear on why this step is required? We're running under the lock, so between the start of this function and now, the only change that could have happened to the concurrency level is that, if it was below the threshhold, it could have gone up by one in step 1 above (in which case there are perforce no non-empty queues). If there was anything queued, then our concurrency limit is full, and the executing requests' afterExecute functions should dequeue the next requests.

What am I missing?

MikeSpreitzer replied:
Step 1 only puts the request in a queue; step 2 can take the request out of that queue and start execution.

What we know at this point, due to the lock covering both steps 1 and 2, is that there is at least one request in the queue.  If we add knowledge about how reconfigurations are handled, we also know that at most one request can be dispatched at this point.  We _could_ add a special case dispatch-at-most-1 function, but I think it is simpler to just re-use the dispatch-as-many-as-possible function.

jktomer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:
```diff
@@ -235,144 +213,136 @@ const (
+	if req.waitStarted {
+		// This can not happen, because the client is forbidden to
+		// call Wait twice on the same request
+		panic(*req)
```
Please add a more descriptive note as well, to save time for someone chasing down the panic caused by some inevitable future bug.

MikeSpreitzer replied:
Are you asking for a more extensive comment, or commentary to go into the argument to `panic`?

jktomer replied:
Commentary for the argument to `panic`. I know someone could follow the stack trace to this line of code and read the comment, but when people are tracking down a panic, any help you can lend is useful (since it's often not just the server that's panicking :wink:)

MikeSpreitzer replied:
done

jktomer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:
```diff
@@ -235,144 +213,136 @@ const (
+	decisionAny := req.decision.GetLocked()
+	qs.syncTimeLocked()
+	decision, isDecision := decisionAny.(requestDecision)
+	if !isDecision {
```
I think it's probably reasonable to just do the unguarded cast and let the compiler-generated panic happen in this case.

MikeSpreitzer replied:
converted to an explicit panic with some identifiers.  More helpful than the compiler-generated panic.

jktomer reviewed (commented):
Still looking this over, but so far I think it looks quite improved. Will read more tomorrow.

logicalhan commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:
```diff
@@ -0,0 +1,124 @@
+	// TODO: decide whether to use the existing metrics, which
+	// categorize according to mutating vs readonly, or make new
+	// metrics because this filter does not pay attention to that
+	// distinction
```
Prima facie, I'd prefer new metrics rather than repurposing existing ones in ways not originally intended. I'll dig into the request metrics a bit more to get a better idea though. 

/sig instrumentation

MikeSpreitzer replied:
Oops, that TODO note is rather outdated.  We have definitely started on using metrics appropriate for this feature.  We certainly need those.  The KEP calls out a design for those.

MikeSpreitzer replied:
Revised the comment.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:
```diff
@@ -0,0 +1,124 @@
+		// Serve the request, but asynchronously, and return from here
+		// as soon as either the request is finished or the context is
+		// canceled.  The logic here is heavily cribbed from the
+		// timeout filter.
```
This filter is not doing all the stuff the timeout filter is doing, it just borrows a bit of plumbing.

MikeSpreitzer replied:
Also, this filter is not adjacent to the timeout filter.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:
```diff
@@ -0,0 +1,158 @@
+	}
+}
+
+var crcHashTable = crc64.MakeTable(crc64.ECMA)
```
Right, simply have not gotten to that yet. Been focused on bigger issues.

MikeSpreitzer replied:
done

MikeSpreitzer committed:
Started making configController test behavioral

Uses public API to inject data.  Still peeking at data structure some,
that still needs revision.

Also responded to review comments.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -606,6 +612,18 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 		}
 	}
 
+	requestManagerConfigControllerHookName := "apiserver/request-management-config-controller"
```
done

MikeSpreitzer committed:
picked some lint

MikeSpreitzer committed:
Switched from CRC32 to SHA256 for hashing flow identifier

MikeSpreitzer commented:
There is an example usage with benchmark of SHA256 in #83796

MikeSpreitzer committed:
Switched config consumer testing to behavioral

Stopped testing data structure of controller.

MikeSpreitzer commented:
The force-push fixes auto-picked nits.

MikeSpreitzer commented:
The force-push to ab08aeb6d29 fixed some more nits.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:
```diff
@@ -0,0 +1,122 @@
+			tooManyRequests(r, w)
+			return
+		}
+		defer afterExecute()
```
wrt @lavalamp's previous suggestion about attaching the name of the matched FS/PL to the response header, it doesnt look easy to get know which FS/PL matched from this encapsulation. probably we can refactor `afterExecute` to return a `ResponseDigest` storing the information so that we attach the header on this defer call

MikeSpreitzer replied:
I think we agreed to address that in a follow-up PR.

lavalamp replied:
yup.

jktomer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/rule.go:
```diff
@@ -0,0 +1,179 @@
+	case fctypesv1a1.SubjectKindUser:
+		return subject.User != nil && (subject.User.Name == fctypesv1a1.NameAll || subject.User.Name == user.GetName())
+	case fctypesv1a1.SubjectKindGroup:
+		return subject.Group != nil && (containsString(subject.Group.Name, user.GetGroups(), fctypesv1a1.NameAll))
```
I think this is still wrong: it will match a request from a user in a group named `*` even if the subject group is not a wildcard, but will not match a user not in a group named `*` if the subject group is a wildcard.

MikeSpreitzer replied:
Oh, right, this call is opposite all the others.  Thanks, will fix.

MikeSpreitzer replied:
Need better testing too.

MikeSpreitzer replied:
There is another odd thing in this vicinity.  Some new options slipped in during the evolution.  Originally the only way to match "any user" was to match on a list of groups that includes both `system:authenticated` and `system:unauthenticated`.  You can see this painful cirumlocution in exhibited configurations.  Now we have two simpler choices: match on username `*` or user group name `*`.  Frankly, I am for these.  Their exclusion seems odd.

MikeSpreitzer replied:
Fixed, after making sure testing catches this bug.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:
```diff
@@ -0,0 +1,186 @@
+// This request filter implements https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+
+type implementation struct {
+	ctl fcfc.Controller
```
This is the result of evolution through a previous design where it seemed imperative to minimize the stuff in the controller package.  Now the difference is even smaller than it was then.  I will think about collapsing these two.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:
```diff
@@ -0,0 +1,186 @@
+	}
+}
+
+var hash32 bool
```
The original code was very aggressive on performance, not doing any allocations.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:
```diff
@@ -0,0 +1,186 @@
+	hash.Write([]byte{1})
+	hash.Write([]byte(fDistinguisher))
+	var sum [8]byte
+	hash.Sum(sum[:])
```
Oh, that's right.  Thanks.

lavalamp replied:
https://play.golang.org/p/s70r4pwLR-V

Surprising behavior, but you should be able to control the alloc like this.

jktomer replied:
Yeah, if the allocation is a problem you can do it that way, though it does require you to know (an upper bound on) the size of the hash which is sort of unfortunate.

jktomer replied:
(by the way, sha256.New and crc64.New both do an explicit heap allocation anyway, so I think you probably don't care.)

MikeSpreitzer replied:
C'mon guys, it is no surprise at all that SHA256 produces a 32 byte hash.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:
```diff
@@ -0,0 +1,186 @@
+
+var crcHashTable = crc64.MakeTable(crc64.ECMA)
+
+func hash32FlowID(fsName, fDistinguisher string) uint64 {
```
32 vs 256 bit hash.
But, as you noted, crc64 exists.  I do not know why the original used crc32.  And yes, the important distinction is CRC vs SHA.

MikeSpreitzer replied:
names fixed

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:
```diff
@@ -235,144 +214,134 @@ const (
+			_ = <-doneCh
+			klog.V(6).Infof("QS(%s): Context of request %#+v %#+v is Done", qs.qCfg.Name, descr1, descr2)
+			qs.cancelWait(req)
+			qs.goroutineDoneOrBlocked()
```
Yes.  The idea is that there is one other increment, at the place that closes `doneCh`.  Now, that actually does not matter when the GoRoutineCounter is a noop --- which is the normal case.  It matters when testing.  Actually, it only matters in a test that relies on correct counting --- that is, a test that uses FakeEventClock::Run --- and cancels a context.  We do not currently have such a test.  But if/when we do, it will do the increment just before canceling the context.

MikeSpreitzer replied:
To make this a little clearer, I am going to put the increment in the one test that cancels a context, even though it does not use FakeEventClock::Run.

MikeSpreitzer replied:
commented here and test updated.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:
```diff
@@ -235,144 +214,134 @@ const (
+	decision, isDecision := decisionAny.(requestDecision)
+	if !isDecision {
+		panic(fmt.Sprintf("QS(%s): Impossible decision %#+v (of type %T) for request %#+v %#+v", qs.qCfg.Name, decisionAny, decisionAny, req.descr1, req.descr2))
+		decision = decisionExecute // yeah, this is a no-op
```
that comment is outdated.  Since this statement is preceded by a panic, this statement will never execute.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,750 @@
+type configController struct {
+	queueSetFactory fq.QueueSetFactory
+
+	// configQueue holds TypedConfigObjectReference values, identifying
```
oops, outdated comment.

MikeSpreitzer replied:
updated

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,750 @@
+	// config holds the configuration after defaulting logic has been applied.
+	// Exempt may be true while there are queues, in the case of a priority
+	// level that recently switched from being non-exempt to exempt and whose
+	// queues are still draining.
```
The comment is outdated; I will fix.

MikeSpreitzer replied:
fixed

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,750 @@
+
+	// number of goroutines between Controller::Match and calling the
+	// returned StartFunction
+	numPending int
```
Because the computation of the flow distinguisher and the hash value is outside in the filter, between Match and the StartFunction.  I deliberately separated the two so that the lock would not be held while doing the SHA256.

jktomer replied:
So a cursory internet search suggests that SHA256 consumes about two nanoseconds per byte hashed; I'm pretty sure this qualifies as a premature optimization.

MikeSpreitzer replied:
I added benchmarks.  Got this result:
```
mjs11:flowcontrol mspreitzus.ibm.com$ go test -bench .
goos: darwin
goarch: amd64
pkg: k8s.io/apiserver/pkg/util/flowcontrol
BenchmarkCRC-8   	19326795	        59.6 ns/op
BenchmarkSHA-8   	 3095313	       388 ns/op
PASS
ok  	k8s.io/apiserver/pkg/util/flowcontrol	3.985s
```

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,750 @@
+	cfgCtl.traceCond = *sync.NewCond(&cfgCtl.lock)
+	klog.V(2).Infof("NewTestableController with serverConcurrencyLimit=%d, requestWaitLimit=%s", serverConcurrencyLimit, requestWaitLimit)
+	cfgCtl.initializeConfigController(informerFactory)
+	cfgCtl.digestConfigObjects(nil, nil)
```
`cfgCtl.digestConfigObjects(..)` _does_ acquire and release the controller's lock.  The acquire will never block, because `cfgCtl` is freshly allocated and there is no concurrent activity at this point.  There are two reasons why using `cfgCtl.digestConfigObjects()` here is a reasonable choice.

(1) There is not currently an existing function that digests config objects without acquiring the lock.

(2) `cfgCtl` is accessed from multiple goroutines and synchronizing its access here follows the simple advice in https://golang.org/ref/mem#tmp_1 .

In fact, in some other languages it is positively necessary to acquire and release the lock before allowing any concurrent access.

In our case, because the access from other goroutines will "happen after" the constructor, we _could_ create and use a function that digests the config objects without acquiring the lock (really this would be a small matter of factoring that functionality out of `lockAndDigestConfigObjects`).  But that seems to me like unnecessary work to make the code more fragile.

jktomer replied:
Sorry, I meant "should we do this in the background, so as not to block the initialization of the apiserver"?

MikeSpreitzer replied:
No, we need to initialize the data structure to include the effects of the mandatory objects.  And testing shows that there is no deadlock here.  But I agree that we should not even tempt that fate.  That will be an easy fix, given way digestConfigObjects is already factored.

MikeSpreitzer replied:
revised and commented

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,750 @@
+// use.  We do this before holding over old priority levels so that
+// requests stop going to those levels and FlowSchemaStatus values
+// reflect this.  This function also adds any missing mandatory
+// FlowSchema objects.  The given objects must all have disstinct
```
typo in "disstinct"

MikeSpreitzer replied:
fixed

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,750 @@
+	// sort into the order to be used for matching
+	sort.Sort(fsSeq)
+
+	// Supply missing mandatory FlowSchemas, in correct position
```
I view it the other way around.  If these are done after the sort, then this code does not have any non-local dependencies on enforcement of restrictions on MatchingPrecedence.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,750 @@
+		if plName == fctypesv1a1.PriorityLevelConfigurationNameExempt && !meal.haveExemptPL || plName == fctypesv1a1.PriorityLevelConfigurationNameCatchAll && !meal.haveCatchAllPL {
+			klog.V(3).Infof("Retaining mandatory priority level %q despite lack of API object", plName)
+		} else {
+			if plState.queues == nil || plState.quiescing && plState.numPending == 0 && plState.queues.IsIdle() {
```
Replacing `plState.quiescing && plState.numPending == 0 && plState.queues.IsIdle()` with `plState.numPending == 0 && plState.queues.IsIdle()` would make this code remove a priority level any time it is empty --- which would be wrong.  We only want to remove a priority level when it is empty _and_ it does not have a corresponding API object.

jktomer replied:
Huh? We've already exited the loop early if the newPLStates[plName] is non-nil (i.e. if the corresponding API object exists). But we *haven't* already set plState.quiescing if the API object does not exist, so this will mean we always have to go through at least two rounds of config changes to drop a PL.

MikeSpreitzer replied:
Oh, right, checking `plState.quiescing` at this point _is_ redundant.  Good catch.

MikeSpreitzer replied:
removed

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,750 @@
+		} else {
+			klog.V(5).Infof("Retaining queues for priority level %q: config=%#+v, concurrencyLimit=%d, quiescing=%v, numPending=%d (shares=%v, shareSum=%v)", plName, fcfmt.Fmt(plState.config), concurrencyLimit, plState.quiescing, plState.numPending, plState.config.Limited.AssuredConcurrencyShares, meal.shareSum)
+		}
+		plState.queues = plState.qsCompleter.Complete(fq.DispatchingConfig{ConcurrencyLimit: concurrencyLimit})
```
This is discussed in the KEP, see https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#reaction-to-configuration-changes .

Now that you raise the question, I see it is not clearly documented in the QueueSet interface.  I will update that.

jktomer replied:
Ah, okay, I think what's happening here is that because any already-queued Requests have pointers to the QueueSet on which they were created (rather than indirecting through their priority level), we can safely drop plState.queues and it won't get GCed until after its queues actually empty out.

In that case, though, why do we bother with the complicated machinery for quiescing plStates? The same logic ought to apply, so long as we make sure to create and enqueue requests from Match().

MikeSpreitzer replied:
Note that when starting from a particular QueueSet X, calling X.BeginConfigChange(), taking the QueueSetCompleter returned from that, then calling its Complete method is defined to return the original QueueSet X (with side effects having happened).  So the line being discussed here is not throwing anything away.  Quiescing is about retaining a priority level, including its QueueSet, while its queues drain after its corresponding PriorityLevelConfiguration object is deleted.

MikeSpreitzer replied:
Updated QueueSet commentary to emphasize this.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,750 @@
+	plState := cfgCtl.priorityLevelStates[plName]
+	if plState != nil && (plState.queues == nil || plState.quiescing && plState.numPending == 0 && plState.queues.IsIdle()) {
+		klog.V(3).Infof("Triggered API Priority and Fairness config reload because priority level %s became idle", plName)
+		cfgCtl.configQueue.Add(0)
```
There are two reasons why it does not.

(1) Because the caller does not hold the lock, it is not guaranteed that the named priority level still exists in the controller's map.  So that is necessary to check here but not in maybeReapLocked.

(2) Also because the caller does not hold the lock, it is not guaranteed that any clue the caller got about the QueueSet being idle is still valid.  So that also has to be checked here and not in maybeReapLocked.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:
```diff
@@ -0,0 +1,122 @@
+			defer func() {
+				err := recover()
+				if err != nil && err != http.ErrAbortHandler {
+					// Same as stdlib http server code. Manually allocate stack
```
I think in at least one other place we actually just send whatever this is through a channel and re-panic, to be sure that the ordinary panic handler gets it. (I don't think I'm asking for a change right now)

jktomer replied:
That actually is what happens here (as well as in the timeout handler). I think the extra stack trace gets logged so that the original panic's trace doesn't get lost; sadly, I don't think Go has a Java-like way to capture a panic's original stack trace in a non-lossy way for repanicking later.

(Even more sadly, the fact that this is also done in the timeout controller, and possibly other places too, means that the ultimate panic is going to have lots of stack traces on it, which kind of sucks.)

MikeSpreitzer replied:
Also in Java it happens that there are multiple printouts of various vintages of a problematic stack.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_filter.go:
```diff
@@ -0,0 +1,186 @@
+	hash.Write([]byte(fDistinguisher))
+	var sum [8]byte
+	hash.Sum(sum[:])
+	return uint64(sum[0])<<0 +
```
Is this better than `import "encoding/binary"` and using e.g. `binary.LittleEndian.Uint64()`?

If we need to optimize this later we can consider the super fast unsafe cast, since byte order makes no difference at all here. For now let's stick to the readable thing.

MikeSpreitzer replied:
Temporary brain slip, I lost track of where to find that.

MikeSpreitzer replied:
fixed

MikeSpreitzer committed:
Responded to some review comments

Fixed hashing.  Added benchmarking for hashing.

Improved testing to catch bug in matching group-type Subject.
Fixed bug.

Fixed some outdated comments.  Added some comments about things that
were not obvious to reviewers.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go:
```diff
@@ -0,0 +1,750 @@
+	haveExemptPL, haveCatchAllPL, haveExemptFS, haveCatchAllFS bool
+
+	// buffered FlowSchema status updates to do.
+	// Do them when the lock is not held, of course.
```
elaborated

MikeSpreitzer commented:
Collapsing the filter and config consumer into one layer and doing the hash with the lock held will reduce the three callbacks to one.

Instead of return; dostuff; callback we could make a dostuff parameter.

MikeSpreitzer committed:
Final tweaks in response to review

Finished updating comments.

Removed an unreachable line after a panic.

Simplified the benchmarks for hashing.

MikeSpreitzer commented:
I simplifed the hashing benchmarks.  Now they look like this:
```
func BenchmarkSHA(b *testing.B) {
	for i := 0; i < b.N; i++ {
		 shaFlowID("supercalifragilistic expialidocious", "the sound of it is something quite atrocious")
	}
}
```

On my 2.9 GHz Core i7 laptop, I get this result:
```
mjs11:flowcontrol mspreitzus.ibm.com$ go test -bench .
goos: darwin
goarch: amd64
pkg: k8s.io/apiserver/pkg/util/flowcontrol
BenchmarkCRC-8   	 4647188	       259 ns/op
BenchmarkSHA-8   	 2010090	       595 ns/op
PASS
ok  	k8s.io/apiserver/pkg/util/flowcontrol	3.370s
```
That's hashing a total of 79 bytes of flow identifier --- 80 bytes total when you include the separator.

MikeSpreitzer commented:
I have responded to all the review comments except those that call for refactoring.  My plan is to make a new PR with the refactored code and close this PR; that way the refactoring does not detach the older PR discussions from the code.

MikeSpreitzer committed:
gofmt fix

MikeSpreitzer commented:
I asked for an escape analysis of the hashing in commit 45b7b46c5af .  Oddly, stuff escaped that I did not expect.
```
mjs11:flowcontrol mspreitzus.ibm.com$ go test -gcflags -m -bench .
...
./apf_filter.go:163:20: inlining call to sha256.New
...
./apf_filter.go:162:16: shaFlowID fsName does not escape
./apf_filter.go:162:24: shaFlowID fDistinguisher does not escape
./apf_filter.go:164:6: moved to heap: sep
./apf_filter.go:168:6: moved to heap: sum
./apf_filter.go:163:20: new(sha256.digest) escapes to heap
./apf_filter.go:163:20: hash.Hash(sha256.d·2) escapes to heap
./apf_filter.go:165:19: ([]byte)(fsName) escapes to heap
./apf_filter.go:167:19: ([]byte)(fDistinguisher) escapes to heap
...
```

This is with Go version 1.13.4

MikeSpreitzer commented:
See #87754 instead
